### PR TITLE
Keras data generation

### PR DESCRIPTION
**Pull request by lapid92**

### CODE REVIEW

File: model_compression_toolkit/data_generation/keras/image_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+class TensorflowCropFlipImagePipeline(BaseImagePipeline):
+    def __init__(self,
+                 output_image_size: Tuple,
+                 extra_pixels: int):
+        """
+        Initialize the TensorflowCropFlipImagePipeline.
+
+        Args:
+            output_image_size (Tuple): The output image size.
+            extra_pixels (int): Extra pixels add to the input image size. Defaults to 0.
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/image_pipeline.py:85]: to add to the...

File: model_compression_toolkit/data_generation/keras/optimization_utils.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+            Optional[tf.Tensor]: The weight tensor of the last linear layer, or None if not found.
+        """
+        # Iterate through the layers in reverse order.
+        for layer in reversed(model.layers):
+            if isinstance(layer, activation_extractor.linear_layers):
+                # Return the weight tensor of the layer.
+                return layer.weights[0] if layer.weights else None
+
+        # If no suitable layer is found, return None.
+        return None
```
Comment (ofirgo) [inline model_compression_toolkit/data_generation/keras/optimization_utils.py:259]: I couldn't find where is the return value of this method being used, but need to make sure this method knows to handle a case where None is returned

File: tests/data_generation_tests/keras/test_keras_data_generation_runner.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        BaseKerasDataGenerationTest(self, image_normalization_type=ImageNormalizationType.EFFICIENTNET).run_test()
+        BaseKerasDataGenerationTest(self, image_normalization_type=ImageNormalizationType.NO_NORMALIZATION).run_test()
+
+    def test_keras_output_loss_types(self):
+        BaseKerasDataGenerationTest(self, output_loss_type=OutputLossType.NONE).run_test()
+        BaseKerasDataGenerationTest(self, output_loss_type=OutputLossType.MIN_MAX_DIFF).run_test()
+        BaseKerasDataGenerationTest(self, output_loss_type=OutputLossType.MARGINAL_MIN_MAX_DIFF).run_test()
+
+
+if __name__ == '__main__':
```
Comment (ofirgo) [inline tests/data_generation_tests/keras/test_keras_data_generation_runner.py:59]: why do we need the main here?

File: model_compression_toolkit/data_generation/common/enums.py [hunk 4]
```diff
@@ -65,9 +65,15 @@ class ImageNormalizationType(EnumBaseClass):
     """
     An enum for choosing the image normalization type.
     TORCHVISION: Normalize the images using torchvision normalization.
+    MOBILENET: Normalize the images using keras mobilenet normalization.
+    EFFICIENTNET: Normalize the images using keras efficientnet normalization.
+    RESNET50: Normalize the images using keras resnet50 normalization.
     NO_NORMALIZATION: Do not apply any normalization.
     """
     TORCHVISION = 'torchvision'
+    MOBILENET = 'mobilenet'
```
[the remark this answers is not in the document]
Comment (lapid92) [reply model_compression_toolkit/data_generation/common/enums.py:74]: It's the normalization from keras.applications.mobilenet.preprocess_input. Do you think it's better to change the name to keras_mobilenet?

[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/data_generation/common/enums.py:74]: Is it just applied for MobileNet? or to other networks as well?
First of all, yes, change the name to include the framework.
Maybe in addition update the documentation to include the details regarding when to use this one and not the other options

File: model_compression_toolkit/data_generation/common/enums.py [hunk 5]
```diff
@@ -97,9 +102,11 @@ class OutputLossType(EnumBaseClass):
     An enum for choosing the output loss type.
     NONE: No output loss is applied.
     MIN_MAX_DIFF: Use min-max difference as the output loss.
+    MARGINAL_MIN_MAX_DIFF: Use marginal min-max difference as the output loss.
     """
     NONE = 'none'
     MIN_MAX_DIFF = 'min_max_diff'
+    MARGINAL_MIN_MAX_DIFF = 'marginal_min_max_diff'
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/data_generation/common/enums.py:105]: Solved offline

File: model_compression_toolkit/data_generation/common/constants.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+IMAGE_INPUT = 'image_input'
+
+# Define a constant for the number of channels in input image.
+NUM_INPUT_CHANNELS = 3
+
+# Default initial learning rate constant.
+DEFAULT_INITIAL_LR = 5
+
+# Default output loss multiplier.
+DEFAULT_OUTPUT_LOSS_MULTIPLIER = 0.0001
```
Comment (ofirgo) [inline model_compression_toolkit/data_generation/common/constants.py:27]: put in pytorch constants

**Review by ofirgo — APPROVED** [review]

File: model_compression_toolkit/data_generation/keras/optimization_functions/scheduler_step_functions.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    def __init__(self,
+                 factor: float = 0.5,  # Factor by which the learning rate will be reduced.
+                 patience: int = 10,  # Number of epochs with no improvement after which learning rate will be reduced.
+                 min_delta: float = 1e-4,  # Minimum change in monitored value to qualify as improvement.
+                 cooldown: int = 0,  # Number of epochs to wait before resuming normal operation after reducing
+                 # learning rate.
+                 min_lr: float = 1e-6,  # Lower bound on the learning rate.
+                 sign_number: int = 4,  # Number of significant digits to consider for comparisons.
+                 optim_lr=None,  # Optimizer learning rate variable (e.g., tf.Variable).
+                 ):
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_functions/scheduler_step_functions.py:35]: add documentation

File: model_compression_toolkit/data_generation/keras/optimization_functions/scheduler_step_functions.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    def _reset(self):
+        """
+        Reset the internal state of the learning rate scheduler.
+        """
+        self.monitor_op = lambda a, b: np.less(a, b - self.min_delta)
+        self.best = np.Inf
+        self.cooldown_counter = 0
+        self.wait = 0
+
+    def on_train_begin(self, logs=None):
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_functions/scheduler_step_functions.py:60]: type hints, missing documentation

File: model_compression_toolkit/data_generation/keras/optimization_utils.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
...
+                                                             scheduler=scheduler,
+                                                             initial_lr=data_generation_config.initial_lr,
+                                                             normalization_mean=self.normalization_mean,
+                                                             normalization_std=self.normalization_std,
+                                                             clip_images=data_generation_config.clip_images,
+                                                             reflection=data_generation_config.reflection,
+                                                             eps=eps)
+
+        # Set the mean axis based on the image granularity
+        self.mean_axis = activation_extractor.mean_axis
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_utils.py:102]: I don't think activation_extractor should have the "mean_axis"

File: model_compression_toolkit/data_generation/keras/optimization_utils.py [hunk 10]
```diff
@@ -0,0 +1,436 @@
...
+            return z
+
+    def get_accumulated_stats_per_layer(self, layer_name: str) -> Tuple[tf.Tensor, tf.Tensor]:
+        """
+        Get the accumulated activation statistics for a layer.
+
+        Args:
+            layer_name (str): the name of the layer.
+
+        Returns: Tuple[tf.Tensor, tf.Tensor]: The averaged activation statistics (mean and second moment) on all the
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_utils.py:176]: go down one line after "Returns" (sorry...)

File: model_compression_toolkit/data_generation/keras/optimization_utils.py [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+        batches for the specified layer.
+        """
+        total_mean, total_second_moment = 0, 0
+        for i_batch in range(self.n_batches):
+            batch_stats_holder = self.all_imgs_stats_holder.batches_stats_holder_list[i_batch]
+            mean, second_moment = self.all_imgs_stats_holder.get_stats(batch_stats_holder=batch_stats_holder,
+                                                                       layer_name=layer_name)
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_utils.py:183]: I don't understand this code. all_images_stats_holder should have that accumulated stats across all batches. You shouldn't do the calculation here. This function is just a "get" function not a "compute" function.

File: model_compression_toolkit/data_generation/keras/optimization_utils.py [hunk 12]
```diff
@@ -0,0 +1,436 @@
...
+        Get the statistics for a given batch and layer.
+
+        Args:
+            batch_stats_holder (BatchStatsHolder): batch stats holder.
+            layer_name (str): the name of the layer.
+
+        Returns:
+            Tuple[tf.Tensor, tf.Tensor]: the mean and second moment for the specified batch and layer.
+        """
+        mean = batch_stats_holder.get_mean(bn_layer_name=layer_name)
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_utils.py:399]: This should give the statistics on all batches not a specific batch. For a specific batch you have "batch_stats_holder"

File: model_compression_toolkit/data_generation/keras/optimization_utils.py [hunk 13]
```diff
@@ -0,0 +1,436 @@
...
+            bn_layer_std = tf.sqrt(bn_layer_var + self.eps)
+            imgs_layer_std = tf.sqrt(imgs_layer_var + self.eps)
+
+            total_bn_loss += layer_weight * self.bn_loss_fn(bn_mean=bn_layer_mean,
+                                                            input_mean=imgs_layer_mean,
+                                                            bn_std=bn_layer_std,
+                                                            input_std=imgs_layer_std)
+        return total_bn_loss
+
+    def compute_output_loss(self,
```
Comment (lior-dikstein) [inline model_compression_toolkit/data_generation/keras/optimization_utils.py:243]: output_loss_multiplier should be outside this function, in the main "data_generation" function

**Review by lior-dikstein — APPROVED** [review]
